File: src/clj/example/vl/goo.clj
(ns example.vl.goo
  (:use [vl shape graph])
  (:import [com.sun.org.apache.bcel.internal Repository]
           [com.sun.org.apache.bcel.internal.classfile Field Method]
           [com.sun.org.apache.bcel.internal.generic ConstantPoolGen MethodGen]))

(def java-class (memoize (fn [classname] (Repository/lookupClass (Class/forName classname)))))
(def constant-pool (memoize (fn [classname] (ConstantPoolGen. (.getConstantPool (java-class classname))))))

(defn java-methods [classname] (.getMethods (java-class classname)))
(defn java-fields [classname] (.getFields (java-class classname)))

(defn is-public? [m] (.isPublic m))
(defn is-static? [m] (.isStatic m))
(defn is-initializer? [m] (= (.getName m) "<init>"))
(defn is-abstract? [m] (.isAbstract m))
(defn is-native? [m] (.isNative m))

(defn instance-fields [classname] (filter is-static? (java-fields classname)))
(defn instance-methods [classname] (filter #(not (or (is-static? %) (is-initializer? %))) (java-methods classname)))

(defn nonvisible-methods [classname] (remove #(is-public? %) (instance-methods classname)))

(defn public-methods [classname] (filter #(is-public? %) (instance-methods classname)))

(defmulti make-member-shape class)
(defmethod make-member-shape Method [m]  (make-shape {:text (.getName m) :x 100 :y 100 :member m}))
(defmethod make-member-shape Field [f]   (make-shape {:text (.getName f) :x 200 :y 150 :member f}))

(defn instruction-list-for [classname method]
  (iterator-seq (.iterator (.getInstructionList (MethodGen. method classname (constant-pool classname))))))

(defn instructions-for [classname method]
  (map #(. % getInstruction) (instruction-list-for classname method)))

(defn is-field-or-method-reference? [inst]
  (some #(isa? % inst) '(GETFIELD PUTFIELD INVOKESPECIAL INVOKEVIRTUAL)))

(defn field-or-method-references [classname method]
  (filter is-field-or-method-reference? (instructions-for classname method)))

(defn inst-for-this-class-p [classname constant-pool]
  (fn [inst] (= classname (.getClassName inst constant-pool))))

(defn uses-from [classname m]
  (filter (inst-for-this-class-p classname (constant-pool classname)) (field-or-method-references classname m)))

(defn edges-from-node [gr n]
  (let [member (:member n)
        classname (:classname gr)
        locator (fn [m] (first (filter #(identical? m (:member %)) (get-nodes gr))))]
    (cond
      (is-abstract? member) '()
      (is-native? member) '()
      (instance? Field member) '()
      true (map locator (uses-from classname member))
      )))

(defn graph-of-class [classname]
  (struct-map directed-graph
    :classname classname
    :fields (instance-fields classname)
    :nonvisible-methods (nonvisible-methods classname)
    :public-methods (public-methods classname)
    :nodes (map make-member-shape (concat (instance-fields classname) (nonvisible-methods classname) (public-methods classname)))
    :neighbors edges-from-node
    ))

File: src/java/goo/src/goo/tool/Goo.java
package goo.tool;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

import com.sun.org.apache.bcel.internal.Repository;
import com.sun.org.apache.bcel.internal.classfile.*;
import com.sun.org.apache.bcel.internal.generic.*;
import com.sun.org.apache.bcel.internal.generic.FieldOrMethod;

public class Goo {
	private JavaClass clazz;
	private List<Edge> nodeEdges = new ArrayList<Edge>();
	private ConstantPoolGen constantPool;

	public Goo(String className)  {
        try {
            clazz = Repository.lookupClass(Class.forName(className));
            if (clazz == null)
                throw new InvalidClassName();
            constantPool = new ConstantPoolGen(clazz.getConstantPool());
            build();
        } catch (ClassNotFoundException e) {
            throw new InvalidClassName();
        }
	}

	void build() {
		Method [] methods = clazz.getMethods();
		if (methods == null)
			return;
		for(Method method : methods) {
			if(!(method.isAbstract() || method.isNative()))
					buildMethodEdges(method);
		}
		strip("<init>");
		strip("$");
	}

    void strip(String filterText) {
		List<Edge> newEdgeList = new ArrayList<Edge>();
		for(Edge edge : nodeEdges) {
			if (edge.containsSubstring(filterText))
				continue;
			newEdgeList.add(edge);
		}
		nodeEdges = newEdgeList;
	}


    InstructionList getInstructionListForMethod(Method method) {
		MethodGen methodGen = new MethodGen(method, clazz.getClassName(), constantPool);
		return methodGen.getInstructionList();
	}

	Set<String> getUsesForMethod(Method method) {
		HashSet<String> uses = new HashSet<String>();
		InstructionList instructions = getInstructionListForMethod(method);
		for(Iterator it = instructions.iterator(); it.hasNext(); ) {
			InstructionHandle handle = (InstructionHandle)it.next();
			Instruction instruction = handle.getInstruction();
			if (!isIncludedAccessInstruction(instruction))
				continue;
			FieldOrMethod accessInstruction = (FieldOrMethod)instruction;
			if (!isInstructionForThisClass(accessInstruction))
				continue;
			uses.add(accessInstruction.getName(constantPool));
		}
		return uses;
	}

	boolean isInstructionForThisClass(FieldOrMethod accessInstruction) {
		return clazz.getClassName().equals(accessInstruction.getClassName(constantPool));
	}

	boolean isIncludedAccessInstruction(Instruction instruction) {
		return instruction instanceof GETFIELD
			|| instruction instanceof PUTFIELD
			|| instruction instanceof INVOKESPECIAL
			|| instruction instanceof INVOKEVIRTUAL;
	}

	void buildMethodEdges(Method method) {
		Set<String> uses = getUsesForMethod(method);
		for(String use : uses) {
			nodeEdges.add(new Edge(method.getName(), use));
		}
	}

        public List<Edge> getEdges() {
		return nodeEdges;
	}

	public List<String> getFieldNodes () {
		List<String> result = new ArrayList<String>();
		Field [] fields = clazz.getFields();
		for(Field field : fields) {
			if (!field.isStatic())
				result.add(field.getName());
		}
		return result;
	}

	public List<String> getNonVisibleMethodNodes() {
		List<String> result = new ArrayList<String>();
		Method [] methods = clazz.getMethods();
		for(Method method : methods) {
			if ((method.isProtected() || method.isPrivate())
					&& !method.isStatic()
					&& !method.getName().equals("<init>"))
				result.add(method.getName());
		}
		return result;
	}

	public List<String> getPublicMethodNodes() {
		List<String> result = new ArrayList<String>();
		Method [] methods = clazz.getMethods();
		for(Method method : methods) {
			if (method.isPublic()
					&& !method.isStatic()
					&& !method.getName().equals("<init>"))
				result.add(method.getName());
		}
		return result;
	}
}
